### PR TITLE
feat: update standardize protocols

### PR DESCRIPTION
**Pull request by godekina**
Hi @lukaspetrasek this PR is a continuation of the merged PR at https://github.com/CarmineOptions/derisk-research/pull/123/commits/cca054223b929abc9c06b1b8e6ee1501f9f623c7 
I have added code to standardize events data for all other protocols in addition to zklend.

kindly review.

### CODE REVIEW

File: notebooks/standardize_events_data.ipynb [hunk 1]
```diff
@@ -22,37 +34,234 @@
...
-   "execution_count": null,
+   "execution_count": 2,
    "id": "f8f05b8e-213b-47ef-9189-486d2e702267",
    "metadata": {},
    "outputs": [],
    "source": [
-    "import pandas as pd\n",
-    "# Define the GCS path to the Parquet file\n",
-    "gcs_path = 'https://storage.googleapis.com/derisk-persistent-state/zklend_data/events_sample.parquet'\n",
+    "# Dictionary of Parquet URLs\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:42]: Pls use the same method for loading data as here: https://github.com/CarmineOptions/derisk-research/blob/master/notebooks/Liquidators_data.ipynb (cell 2). No need to use `gcsfs`.

File: notebooks/standardize_events_data.ipynb [hunk 2]
```diff
@@ -22,37 +34,234 @@
...
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
    "source": [
     "# taking a look at the entire second row\n",
     "pd.set_option('display.max_colwidth', None)\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:267]: Move this setup to the first cell.

File: notebooks/standardize_events_data.ipynb [hunk 3]
```diff
@@ -70,15 +279,13 @@
...
+   "execution_count": 5,
    "id": "03358fb0-dd7b-4c13-a57a-4703a2622ecd",
    "metadata": {},
    "outputs": [],
    "source": [
     "def decode_byte_string(value):\n",
-    "    if isinstance(value, bytes):\n",
-    "        return value.decode(\"utf-8\")\n",
-    "    return value"
+    "    return value.decode(\"utf-8\")  "
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:288]: Do you need a function for this? Isn't it simpler to use the `decode` method directly?

File: notebooks/standardize_events_data.ipynb [hunk 4]
```diff
@@ -90,90 +297,914 @@
...
+    "                'token': data_list[2],\n",
+    "                'raw_amount': hex_to_int(data_list[3])\n",
+    "            })\n",
+    "        elif row['key_name'] == 'AccumulatorsSync':\n",
+    "            return pd.Series({\n",
+    "                'token': data_list[0],\n",
+    "                'lending_accumulator': data_list[1],\n",
+    "                'debt_accumulator': data_list[2]\n",
+    "            })\n",
+    "        else:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:383]: What are the other key names that we ignore?

File: notebooks/standardize_events_data.ipynb [hunk 5]
```diff
@@ -90,90 +297,914 @@
...
+    "            })\n",
+    "        elif row['key_name'] == 'InterestStateUpdated':\n",
+    "            return pd.Series({\n",
+    "                'debt_token': data_list[0],\n",
+    "                'lending_rate': data_list[1],\n",
+    "                'borrow_rate': data_list[3],\n",
+    "                'lend_index': data_list[5],\n",
+    "                'borrow_index': data_list[7]\n",
+    "            })\n",
+    "        else:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:455]: What are the other key names that we ignore? For example, `Burn` events are missing.

File: notebooks/standardize_events_data.ipynb [hunk 6]
```diff
@@ -90,90 +297,914 @@
...
+    "    \n",
+    "    def hex_to_int(hex_str):\n",
+    "        return int(hex_str, 16)\n",
+    "    \n",
+    "    try:\n",
+    "        # Decode the bytes to a string and convert the string to an actual list\n",
+    "        data_str = row['data'].decode('utf-8')\n",
+    "        data_list = eval(data_str)\n",
+    "   \n",
+    "        if row['key_name'] == 'Mint':\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:435]: From this, we don't know whether the `Mint` event represents a deposit or borrowing event. Pls add that information by mapping `token` on the type of token defined here: https://github.com/CarmineOptions/derisk-research/blob/master/src/nostra_alpha.py#L36.

For example, `Mint` events for `non_interest_bearing_collateral` or `interest_bearing_collateral` are deposit events, for `debt` they're borrowing events, etc. Map the event type names on those used for zkLend.

File: notebooks/standardize_events_data.ipynb [hunk 7]
```diff
@@ -90,90 +297,914 @@
...
+    "        elif row['key_name'] == 'liquidated_event':\n",
+    "            return pd.Series({\n",
+    "                'loan_id': hex_to_int(data_list[0]),\n",
+    "                'user': data_list[1],\n",
+    "                'debt_token': data_list[2],\n",
+    "                'borrowed_collateral_token': data_list[6],\n",
+    "                'borrowed_collateral_face_amount': hex_to_int(data_list[7]),\n",
+    "                'debt_category': hex_to_int(data_list[10])\n",
+    "            })\n",
+    "        else:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:569]: What are the other key names that we ignore?

File: notebooks/standardize_events_data.ipynb [hunk 8]
```diff
@@ -90,90 +297,914 @@
...
+    "                'new_original_collateral_token': data_list[25],\n",
+    "                'new_original_collateral_face_amount': hex_to_int(data_list[26])\n",
+    "            })\n",
+    "        elif row['key_name'] == 'loan_transfered':\n",
+    "            return pd.Series({\n",
+    "                'loan_id': hex_to_int(data_list[0]),\n",
+    "                'old_user': data_list[1],\n",
+    "                'new_user': data_list[2]\n",
+    "            })\n",
+    "        else:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:971]: What are the other key names that we ignore?

**Review by lukaspetrasek — CHANGES_REQUESTED** [review]

**Review by lukaspetrasek — CHANGES_REQUESTED** [review]

File: notebooks/standardize_events_data.ipynb [hunk 9]
```diff
@@ -282,7 +282,7 @@
    "id": "9bde802d-8d13-48d3-a099-168f29c87400",
    "metadata": {},
    "source": [
-    "2. Define a Function to Transform Each Row\n",
+    "## 2. Define a Function to Transform Each Row\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:285]:
```suggestion
    "## 2. Define a function to transform each row\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 10]
```diff
@@ -370,26 +381,26 @@
     "                'debt_accumulator': data_list[2]\n",
     "            })\n",
     "        else:\n",
-    "            return pd.Series()\n",
+    "            raise ValueError(\"Unrecognized key_name\") # Raising an error for unrecognized key_name\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:384]: Print the key name that was not recognized when raising the error.

File: notebooks/standardize_events_data.ipynb [hunk 11]
```diff
@@ -370,26 +381,26 @@
...
     "    except Exception as e:\n",
     "        # Handle any parsing errors\n",
-    "        return pd.Series()\n",
+    "        raise RuntimeError(f\"Failed to extract data from row: {str(e)}\") from e\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:387]:
```suggestion
    "        raise ValueError(f\"Failed to extract data from row: {str(e)}\") from e\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 12]
```diff
@@ -449,21 +459,18 @@
     "    }\n",
     "}\n",
     "\n",
-    "def get_event_type(token_address, key_name):\n",
-    "    token_type = TOKEN_MAPPING.get(token_address.lower())\n",
+    "def get_nostra_alpha_event_type(token_address: str, key_name: str) ->str:\n",
+    "    token_type = NOSTRA_ALPHA_TOKENS_TO_TOKEN_TYPES.get(token_address.lower())\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:463]: Why `.lower()`? That shouldn't make any difference.

File: notebooks/standardize_events_data.ipynb [hunk 13]
```diff
@@ -449,21 +459,18 @@
...
     "    if token_type:\n",
-    "        return EVENT_MAPPING.get(token_type, {}).get(key_name)\n",
-    "    return 'interest rate updated'  # Default to 'interest rate updated' if not matched\n",
+    "        return NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES.get(token_type, {}).get(key_name)\n",
+    "    return 'interest_rate_updated'  # Default to 'interest rate updated' if not matched\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:466]: The default should be raising an error (if no match is found). An `interest_rate_updated` should be returned when `token_address` is `0x03d39f7248fb2bfb960275746470f7fb470317350ad8656249ec66067559e892` and `key_name` is `InterestStateUpdated`.

File: notebooks/standardize_events_data.ipynb [hunk 14]
```diff
@@ -482,26 +489,26 @@
...
     "        else:\n",
-    "            return pd.Series()  # Return an empty series for unmatched events\n",
+    "            raise ValueError(\"Unrecognized key_name\") # Raising an error for unrecognized key_name\n",
     "    except Exception as e:\n",
     "        # Handle any parsing errors\n",
-    "        return pd.Series()  # Return an empty series in case of error\n",
+    "        raise RuntimeError(f\"Failed to extract data from row: {str(e)}\") from e\n",
     "\n",
     "# Filter the dataset for 'nostra_alpha' protocol\n",
     "nostra_alpha_data = df_events[df_events['Protocol'] == 'nostra_alpha']\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:498]:
```suggestion
    "nostra_alpha_events = df_events[df_events['Protocol'] == 'nostra_alpha']\n",
```
In general, it's best to be more descriptive whenever you can. `Data` is very general, `events` is specific.

File: notebooks/standardize_events_data.ipynb [hunk 15]
```diff
@@ -1071,27 +1070,24 @@
...
-    "            return pd.Series()\n",
+    "       else:\n",
+    "            raise ValueError(\"Unrecognized key_name\") # Raising an error for unrecognized key_name\n",
     "    except Exception as e:\n",
     "        # Handle any parsing errors\n",
-    "        return pd.Series()\n",
+    "        raise RuntimeError(f\"Failed to extract data from row: {str(e)}\") from e\n",
     "\n",
     "# Filter the dataset for 'hashstack_v1' protocol\n",
     "hashstack_v1_data = df_events[df_events['Protocol'] == 'hashstack_v1']\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:1080]:
```suggestion
    "hashstack_v1_events = df_events[df_events['Protocol'] == 'hashstack_v1']\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 16]
```diff
@@ -370,26 +381,26 @@
...
-    "            return pd.Series()\n",
+    "            raise ValueError(\"Unrecognized key_name\") # Raising an error for unrecognized key_name\n",
     "    except Exception as e:\n",
     "        # Handle any parsing errors\n",
-    "        return pd.Series()\n",
+    "        raise RuntimeError(f\"Failed to extract data from row: {str(e)}\") from e\n",
     "\n",
     "# Filter the dataset for 'zklend' protocol\n",
-    "zk_data = df_events[df_events['Protocol'] == 'zklend']\n",
+    "zklend_events = df_events[df_events['Protocol'] == 'zklend']\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:390]:
```suggestion
    "zklend_events = events[events['Protocol'] == 'zklend']\n",
```
No need to put `df` (i.e. the type of the variable) in the name.

File: notebooks/standardize_events_data.ipynb [hunk 17]
```diff
@@ -449,21 +459,18 @@
...
-    "        return EVENT_MAPPING.get(token_type, {}).get(key_name)\n",
-    "    return 'interest rate updated'  # Default to 'interest rate updated' if not matched\n",
+    "        return NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES.get(token_type, {}).get(key_name)\n",
+    "    return 'interest_rate_updated'  # Default to 'interest rate updated' if not matched\n",
     "\n",
-    "def extract_nostra_alpha(row):\n",
-    "    def hex_to_int(hex_str):\n",
-    "        return int(hex_str, 16)\n",
-    "    \n",
+    "def extract_nostra_alpha(row: pd.Series) -> pd.Series:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:468]:
```suggestion
    "def standardize_nostra_alpha_event(row: pd.Series) -> pd.Series:\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 18]
```diff
@@ -1071,27 +1070,24 @@
...
     "hashstack_v1_data = df_events[df_events['Protocol'] == 'hashstack_v1']\n",
     "\n",
     "# Apply the extraction function to each row\n",
-    "extracted_elements = hashstack_v1_data.apply(extract_hashstack_v1, axis=1)\n",
+    "standardized_hashstack_v1_events = hashstack_v1_data.apply(standardize_hashstack_v1_event, axis=1)\n",
     "\n",
     "# Concatenate the extracted elements with the original dataframe\n",
-    "standard_hashstack_v1 = pd.concat([hashstack_v1_data[['block_hash', 'block_number', 'transaction_hash', 'timestamp', 'key_name']], extracted_elements], axis=1)\n",
-    "\n",
-    "# Display the resulting dataframe\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb]: Why remove the display here and not in the case of other protocols?

File: notebooks/standardize_events_data.ipynb [hunk 19]
```diff
@@ -687,26 +690,26 @@
...
     "        # Handle any parsing errors\n",
-    "        return pd.Series()\n",
+    "        raise RuntimeError(f\"Failed to extract data from row: {str(e)}\") from e\n",
     "\n",
     "# Filter the dataset for 'hashstack_v0' protocol\n",
     "hashstack_v0_data = df_events[df_events['Protocol'] == 'hashstack_v0']\n",
     "\n",
     "# Apply the extraction function to each row\n",
-    "extracted_elements = hashstack_v0_data.apply(extract_hashstack_v0, axis=1)\n",
+    "standardized_hashstack_v0_events = hashstack_v0_data.apply(standardize_hashstack_v0_event, axis=1)\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:702]: I'm gettings `RuntimeError: Failed to extract data from row: list index out of range` here.

File: notebooks/standardize_events_data.ipynb [hunk 20]
```diff
@@ -1071,27 +1070,24 @@
...
     "        # Handle any parsing errors\n",
-    "        return pd.Series()\n",
+    "        raise RuntimeError(f\"Failed to extract data from row: {str(e)}\") from e\n",
     "\n",
     "# Filter the dataset for 'hashstack_v1' protocol\n",
     "hashstack_v1_data = df_events[df_events['Protocol'] == 'hashstack_v1']\n",
     "\n",
     "# Apply the extraction function to each row\n",
-    "extracted_elements = hashstack_v1_data.apply(extract_hashstack_v1, axis=1)\n",
+    "standardized_hashstack_v1_events = hashstack_v1_data.apply(standardize_hashstack_v1_event, axis=1)\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:1083]: I'm getting `RuntimeError: Failed to extract data from row: Unrecognized key_name` here.

File: notebooks/standardize_events_data.ipynb [hunk 21]
```diff
@@ -1071,27 +1070,24 @@
     "                'receiver': data_list[2],\n",
     "                'timestamp': data_list[3]\n",
     "            })\n",
-    "        else:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb]: Fix the indent, this throws a SyntaxError now.

File: notebooks/standardize_events_data.ipynb [hunk 22]
```diff
@@ -1071,27 +1070,24 @@
...
+    "standardized_hashstack_v1_events = pd.concat([hashstack_v1_data[['block_hash', 'block_number', 'transaction_hash', 'timestamp', 'key_name']], standardized_hashstack_v1_events], axis=1)\n",
     "\n",
     "# Uncomment the following below to Save the resulting dataframe to a new CSV file\n",
-    "# output_file_path = 'extracted_hashstack_v1_events.csv'\n",
-    "# standard_hashstack_v1.to_csv(output_file_path, index=False)"
+    "# output_file_path = 'standardized_hashstack_v1_events.csv'\n",
+    "# standardized_hashstack_v1_events.to_csv(output_file_path, index=False)"
    ]
   },
   {
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:1093]: The code for Nostra Mainnet should be adjusted according to the comments above as well. Moreover, you need to copy the setting for `NOSTRA_MAINNET_TOKENS_TO_TOKEN_TYPES` and `NOSTRA_MAINNET_TOKEN_TYPES_TO_EVENT_TYPES` from `src`.

**Comment by godekina** [comment]
@lukaspetrasek all changes requested were made to the notebook, kindly review: 
https://github.com/CarmineOptions/derisk-research/pull/142/commits/6308af629567516188ac87e42dd3aea7f38ce602

File: notebooks/standardize_events_data.ipynb [hunk 23]
```diff
@@ -31,12 +31,12 @@
    "id": "f8f758bf-ec63-4d84-a658-d0461cc6e086",
    "metadata": {},
    "source": [
-    "## 1.  Load event data from Paraquet file"
+    "## 1.  Load event data  the paraquet file"
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:34]:
```suggestion
    "## 1.  Load event data from the paraquet file"
```

File: notebooks/standardize_events_data.ipynb [hunk 24]
```diff
@@ -268,21 +268,21 @@
...
+    "events.head()"
    ]
   },
   {
    "cell_type": "markdown",
    "id": "9bde802d-8d13-48d3-a099-168f29c87400",
    "metadata": {},
    "source": [
-    "## 2. Define a Function to Transform Each Row\n",
+    "## 2. Define a function to transform Each Row\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:285]:
```suggestion
    "## 2. Define a function to transform each row\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 25]
```diff
@@ -2,179 +2,2050 @@
...
+   ],
    "source": [
-    "# taking a look at the entire second row\n",
-    "pd.set_option('display.max_colwidth', None)\n",
-    "df.iloc[1,:]"
+    "# Define a function to decode bytes, convert to list, and extract specific elements based on the event\n",
+    "def hex_to_int(hex_str: str) -> int:\n",
+    "        return int(hex_str, 16)\n",
+    "\n",
+    "def extract_zklend(row: pd.Series) -> pd.Series:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:505]:
```suggestion
    "def standardize_zklend_event(row: pd.Series) -> pd.Series:\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 26]
```diff
@@ -419,13 +596,203 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 5,
+   "execution_count": 19,
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:599]: Move `Nostra Mainnet` right below `Nostra Alpha`.

File: notebooks/standardize_events_data.ipynb [hunk 27]
```diff
@@ -419,13 +596,203 @@
...
+     },
+     "execution_count": 19,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
    "source": [
     "# Define the mapping for token addresses and their types\n",
-    "NOSTRA_ALPHA_TOKENS_TO_TOKEN_TYPES = {\n",
+    "NOSTRA_ALPHA_TOKENS_TO_TOKEN_TYPES :dict[str, str] =  { \n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:795]:
```suggestion
    "NOSTRA_ALPHA_TOKENS_TO_TOKEN_TYPES: dict[str, str] =  { \n",
```

File: notebooks/standardize_events_data.ipynb [hunk 28]
```diff
@@ -443,8 +810,9 @@
     "    '0x075b0d87aca8dee25df35cdc39a82b406168fa23a76fc3f03abbfdc6620bb6d7': 'debt'\n",
     "}\n",
     "\n",
+    "NOSTRA_ALPHA_INTEREST_RATE_MODEL_ADDRESS: str = '0x03d39f7248fb2bfb960275746470f7fb470317350ad8656249ec66067559e892'\n",
     "# Define the event type mapping based on the token type and key name\n",
-    "NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES = {\n",
+    "NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES : dict[str, str] = {\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:815]:
```suggestion
    "NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES: dict[str, str] = {\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 29]
```diff
@@ -460,10 +828,17 @@
...
     "    if token_type:\n",
-    "        return NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES.get(token_type, {}).get(key_name)\n",
-    "    return 'interest_rate_updated'  # Default to 'interest rate updated' if not matched\n",
+    "        event_type = NOSTRA_ALPHA_TOKEN_TYPES_TO_EVENT_TYPES.get(token_type, {}).get(key_name)\n",
+    "        if event_type:\n",
+    "            return event_type\n",
+    "    # If no match is found, raise an error\n",
+    "    raise ValueError(f\"Unrecognized event type for token address {token_address} and key name {key_name}\")\n",
     "\n",
     "def extract_nostra_alpha(row: pd.Series) -> pd.Series:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:843]:
```suggestion
    "def standardize_nostra_alpha_event(row: pd.Series) -> pd.Series:\n",
```

File: notebooks/standardize_events_data.ipynb [hunk 30]
```diff
@@ -690,22 +1309,22 @@
     "                'timestamp': data_list[15]\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:1309]: There's a lot of variables that we actually don't need. Check src.hashstack_v0.py to see what's needed and what's not needed. Ignore variables which are not needed. Holds for Hashstack V1, too.

File: notebooks/standardize_events_data.ipynb [hunk 31]
```diff
@@ -1296,28 +1918,134 @@
...
+    "\n",
+    "    token_type = NOSTRA_MAINNET_TOKENS_TO_TOKEN_TYPES.get(token_address)\n",
+    "    if token_type:\n",
+    "        event_type = NOSTRA_MAINNET_TOKEN_TYPES_TO_EVENT_TYPES.get(token_type, {}).get(key_name)\n",
+    "        if event_type:\n",
+    "            return event_type\n",
+    "    # If no match is found, raise an error\n",
+    "    raise ValueError(f\"Unrecognized event type for token address {token_address} and key name {key_name}\")\n",
+    "\n",
+    "def extract_nostra_mainnet(row: pd.Series) -> pd.Series:\n",
```
Comment (lukaspetrasek) [inline notebooks/standardize_events_data.ipynb:1996]:
```suggestion
    "def standardize_nostra_mainnet_event(row: pd.Series) -> pd.Series:\n",
```

**Pull request merged**